compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+        };
+
+        // Get the callable.
+        let global = self
+            .package_store
+            .get_global(*store_item_id)
+            .expect("global not present");
+        let Global::Callable(callable_decl) = global else {
+            // Instruction generation for UDTs is not supported.
+            panic!("global is not a callable");
+        };
+        (*store_item_id, *functor_app, callable_decl.clone())
+    }
+
+    fn eval_classical_expr(&mut self, expr_id: ExprId) {
+        let current_package_id = self.get_current_package_id();
+        let store_expr_id = StoreExprId::from((current_package_id, expr_id));
+        let expr = self.package_store.get_expr(store_expr_id);
+        let scope_exec_graph = self.get_current_scope_exec_graph().clone();
+        let scope = self.eval_context.get_current_scope_mut();
+        let exec_graph = exec_graph_section(&scope_exec_graph, expr.exec_graph_range.clone());
+        let mut out = Vec::new();
+        let mut receiver = GenericReceiver::new(&mut out);
+        let mut state = State::new(current_package_id, exec_graph, None);
+        let eval_result = state.eval(
+            self.package_store,
+            &mut scope.env,
+            &mut self.backend,
+            &mut receiver,
+            &[],
+            StepAction::Continue,
+        );
+        match eval_result {
+            Ok(step_result) => {
+                let StepResult::Return(value) = step_result else {
+                    panic!("evaluating a classical expression should always return a value");
+                };
+                self.eval_context
+                    .get_current_scope_mut()
+                    .insert_expr_value(expr_id, value);
+            }
+            Err((error, _)) => self.error = Some(Error::EvaluationFailed(error)),
+        };
+    }
+
+    fn eval_classical_stmt(&mut self, stmt_id: StmtId) {
+        let current_package_id = self.get_current_package_id();
+        let store_stmt_id = StoreStmtId::from((current_package_id, stmt_id));
+        let stmt = self.package_store.get_stmt(store_stmt_id);
+        let scope_exec_graph = self.get_current_scope_exec_graph().clone();
+        let scope = self.eval_context.get_current_scope_mut();
+        let exec_graph = exec_graph_section(&scope_exec_graph, stmt.exec_graph_range.clone());
+        let mut out = Vec::new();
+        let mut receiver = GenericReceiver::new(&mut out);
+        let mut state = State::new(current_package_id, exec_graph, None);
+        _ = state.eval(
+            self.package_store,
+            &mut scope.env,
+            &mut self.backend,
+            &mut receiver,
+            &[],
+            StepAction::Continue,
+        );
+    }
+
+    fn generate_expr_call(&mut self, callee_expr_id: ExprId, args_expr_id: ExprId) {
+        let (store_item_id, functor_app, callable_decl) = self.eval_callee_expr(callee_expr_id);
+
+        // We generate instructions differently depending on whether
```

Comment:
whoops, looks like the comment got cut off.

Reply:
It seems like my brain `panicked` in this one 😆 
Updated with a complete comment.